src/routine.ts

```diff
@@ -0,0 +1,291 @@
+       * const bigquery = new BigQuery();
+       * const dataset = bigquery.dataset('my-dataset');
+       * const routine = dataset.routine('my_routine');
+       *
+       * routine.getMetadata((err, metadata, apiResponse) => {});
+       *
+       * @example <caption>If the callback is omitted a Promise will be returned</caption>
+       * const [metadata, apiResponse] = await routine.getMetadata();
+       */
+      getMetadata: true,
+
+      /**
+       * @callback SetRoutineMetadataCallback
+       * @param {?Error} err Request error, if any.
+       * @param {object} metadata The routine metadata.
+       * @param {object} apiResponse The full API response.
+       */
+      /**
+       * @typedef {array} SetRoutineMetadataResponse
+       * @property {object} 0 The routine metadata.
+       * @property {object} 1 The full API response.
+       */
+      /**
+       * Update a routine.
+       *
+       * @see [Routines: update API Documentation]{@link https://cloud.google.com/bigquery/docs/reference/rest/v2/routines/update}
+       *
+       * @method Routine#setMetadata
+       * @param {object} metadata A [routine resource object]{@link https://cloud.google.com/bigquery/docs/reference/rest/v2/routines#Routine}.
+       * @param {SetRoutineMetadataCallback} [callback] The callback function.
+       * @returns {Promise<SetRoutineMetadataResponse>}
+       *
+       * @example
+       * const {BigQuery} = require('@google-cloud/bigquery');
+       * const bigquery = new BigQuery();
+       * const dataset = bigquery.dataset('my-dataset');
+       * const routine = dataset.routine('my_routine');
+       *
+       * const updates = {
+       *   description: 'The perfect description!'
+       * };
+       *
+       * routine.setMetadata(updates, (err, metadata, apiResponse) => {});
+       *
+       * @example <caption>If the callback is omitted a Promise will be returned</caption>
+       * const [metadata, apiResponse] = await routine.setMetadata(updates);
+       */
+      setMetadata: {
+        reqOpts: {
+          method: 'PUT',
+        },
+      },
+    };
+
+    super({
+      parent: dataset,
+      baseUrl: '/routines',
+      id,
+      methods,
+      createMethod: dataset.createRoutine.bind(dataset, id),
+    });
+  }
+
+  setMetadata(metadata: RoutineMetadata): Promise<common.SetMetadataResponse>;
+  setMetadata(
+    metadata: RoutineMetadata,
+    callback: common.ResponseCallback
+  ): void;
+  setMetadata(
+    metadata: RoutineMetadata,
+    callback?: common.ResponseCallback
+  ): void | Promise<common.SetMetadataResponse> {
+    // per the python client, it would appear that in order to update a routine
+    // you need to send the routine in its entirety, not just the updated fields
+    this.getMetadata((err: Error | null, fullMetadata: RoutineMetadata) => {
+      if (err) {
+        callback!(err);
+        return;
+      }
+
+      const updatedMetadata = extend(true, {}, metadata, fullMetadata);
```

Comment:
Shouldn't the user-provided metadata win?

Reply:
Oof, yes, good catch.